File: wordladder/wordladder.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        int res = 0;
        unordered_set<string> wordSet(wordList.begin(), wordList.end());
        if (!wordSet.count(endWord) || beginWord == endWord)
            return res;
        unordered_set<string> start, end;
        start.insert(beginWord);
        end.insert(endWord);
        while (!start.empty() && !end.empty()) {
            if (start.size() > end.size())
                start.swap(end);
            res++;
            unordered_set<string> next;
            for (string word : start) {
                for (int i = 0; i < word.length(); i++) {
                    for (int j = 0; j < 26; j++) {
                        char c = word[i];
                        if (c == 'a' + j)
                            continue;
                        word[i] = 'a' + j;
                        if (!wordSet.count(word)) {
                            word[i] = c;
                            continue;
                        }
                        if (end.count(word))
                            return res + 1;
                        next.insert(word);
                        word[i] = c;
                    }
                } // for
                wordSet.erase(word);
            }// for
            start = next;
        }//while
        return 0;
    }
};

int main()
{
    Solution s;
    string beginWord = "hit";
    string endWord = "cog";
    vector<string> wordList = {"hot", "dot", "lot", "log", "cog"};
    cout << s.ladderLength(beginWord, endWord, wordList) << endl;
    return 0;
}

File: wordLadder/wordladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>();
        for (String word : wordList) wordSet.add(word);
        int res = 0;
        Queue<String> q = new LinkedList<>();
        q.offer(beginWord);
        while (!q.isEmpty()) {
            int q_size = q.size();
            res++;
            for (int i = 0; i < q_size; i++) {
                String curWord = q.poll();
                for (int index = 0; index < curWord.length(); index++) {
                    char[] curWordChars = curWord.toCharArray();
                    for (char c = 'a'; c <= 'z'; c++) {
                        if (c == curWordChars[index]) continue;
                        curWordChars[index] = c;
                        String nextWord = new String(curWordChars);
                        if (!wordSet.contains(nextWord)) continue;
                        if (nextWord.equals(endWord)) return res + 1;
                        wordSet.remove(nextWord);
                        q.offer(nextWord);
                    }
                }
            }
        }
        return 0;
    }
}
